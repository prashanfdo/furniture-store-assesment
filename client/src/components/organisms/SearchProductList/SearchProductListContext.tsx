import constate from "constate";
import { useQuery, gql } from "@apollo/client";
import { Article } from "types";
type GetSearchProductListResponse = {
  searchData: {
    total: number;
    articles: Article[];
  };
};
const GET_PRODUCT_LIST = gql`
  query GetSearchProductList($query: String!, $locale: Locale!) {
    searchData: search(locale: $locale, query: $query) {
      ... on SearchResultArticles {
        total
        articles {
          name
          variantName
          images(format: WEBP, maxWidth: 200, maxHeight: 200, limit: 4) {
            path
            type
          }
          prices {
            currency
            regular {
              value
            }
            special {
              value
              discount
            }
          }
        }
      }
    }
  }
`;

type SearchProductListContextProviderProps = {
  searchText: string;
  page: number;
};
const [SearchProductListContextProvider, useSearchProductListContext] = constate(
  ({ searchText, page }: SearchProductListContextProviderProps) => {
    const { loading, error, data } = useQuery<GetSearchProductListResponse>(GET_PRODUCT_LIST, {
      variables: { query: searchText, locale: "de_DE" },
    });
    console.log(data?.searchData);
    return {
      searchText,
      page,
      loading,
      error,
      data: data?.searchData,
    };
  }
);

export { SearchProductListContextProvider, useSearchProductListContext };
