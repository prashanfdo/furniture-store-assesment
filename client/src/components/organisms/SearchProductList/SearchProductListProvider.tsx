import constate from "constate";
import useGetSearchProductList from "queries/useGetSearchProductList";

type SearchProductListContextProviderProps = {
  searchText: string;
  page: number;
};
const [SearchProductListProvider, useSearchProductListContext] = constate(
  ({ searchText, page }: SearchProductListContextProviderProps) => {
    const { loading, error, data } = useGetSearchProductList(searchText);
    return {
      searchText,
      page,
      loading,
      error,
      data: data?.searchData,
    };
  },
);

export { useSearchProductListContext };

export default SearchProductListProvider;
