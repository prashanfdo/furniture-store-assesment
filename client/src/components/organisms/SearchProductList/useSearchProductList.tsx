import useSearchProductListQuery from "queries/useSearchProductListQuery";

const useSearchProductList = (searchText: string, page: number) => {
  const { loading, error, data } = useSearchProductListQuery(searchText);
  return {
    loading,
    error,
    data: data?.searchData,
  };
};

export default useSearchProductList;
