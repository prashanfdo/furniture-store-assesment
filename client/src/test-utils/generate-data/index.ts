export * from "./article";
export * from "./category";
export * from "./search";
