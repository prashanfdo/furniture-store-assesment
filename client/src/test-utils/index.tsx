import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
