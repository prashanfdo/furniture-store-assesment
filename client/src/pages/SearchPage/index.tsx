import SearchProductList from "components/organisms/SearchProductList";
import { useSearchParams } from "react-router-dom";

export default function SearchPage() {
  const [searchParams] = useSearchParams();
  const searchText = searchParams.get("q") || undefined;

  if (!searchText) {
    return null;
  }

  return (
    <>
      <SearchProductList searchText={searchText} page={1} />
    </>
  );
}
