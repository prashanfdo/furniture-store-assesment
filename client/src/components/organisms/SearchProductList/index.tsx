import { ArticleListHeader } from "components/atoms";
import { ArticleGrid } from "components/molecules";
import useSearchProductList from "./useSearchProductList";

type SearchProductListProps = {
  searchText: string;
  page: number;
};
const SearchProductList: React.FC<SearchProductListProps> = ({ searchText, page }: SearchProductListProps) => {
  const { data } = useSearchProductList(searchText, page);

  if (data === undefined) {
    return null;
  }
  return (
    <div className="grid grid-rows-[auto_1fr] py-2 px-8">
      <ArticleListHeader name={`Search Results for "${searchText}"`} articleCount={data.total} />
      <ArticleGrid products={data.articles as any} />
    </div>
  );
};

export default SearchProductList;
