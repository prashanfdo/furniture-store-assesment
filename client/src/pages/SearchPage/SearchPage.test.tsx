import { render, screen } from "test-utils";
import SearchPage from "./index";
import SearchProductList from "components/organisms/SearchProductList";
import { useSearchParams } from "react-router-dom";

jest.mock(
  "components/organisms/SearchProductList",
  () =>
    ({ searchText, page }: React.ComponentProps<typeof SearchProductList>) =>
      <div data-testid="search-product-list" data-search-text={searchText} data-page={page} />,
);
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useSearchParams: jest.fn(),
}));
const mockUseSearchParams = useSearchParams as jest.MockedFunction<typeof useSearchParams>;

describe("SearchPage", () => {
  it("should render", () => {
    mockUseSearchParams.mockReturnValue([new URLSearchParams("q=test"), jest.fn()]);
    const { container } = render(<SearchPage />);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("search-product-list")).toHaveAttribute("data-search-text", "test");
    expect(screen.getByTestId("search-product-list")).toHaveAttribute("data-page", "1");
  });

  it("should render when no search text", () => {
    mockUseSearchParams.mockReturnValue([new URLSearchParams(""), jest.fn()]);
    const { container } = render(<SearchPage />);
    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("search-product-list")).not.toBeInTheDocument();
  });
});
