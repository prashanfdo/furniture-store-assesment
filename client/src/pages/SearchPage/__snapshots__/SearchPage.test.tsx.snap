// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should render 1`] = `
<div>
  <div
    data-page="1"
    data-search-text="test"
    data-testid="search-product-list"
  />
</div>
`;

exports[`SearchPage should render when no search text 1`] = `<div />`;
